StaticAnalyzer/views/android/find.py

```diff
@@ -7,64 +7,75 @@
 from django.utils.html import escape
 
 from MobSF.utils import (
     PrintException
 )
 
-def run(request):
+def run(request, is_api=False):
     """Find in source files."""
     try:
-        match = re.match('^[0-9a-f]{32}$', request.POST['md5'])
-        if match:
-            md5 = request.POST['md5']
-            query = request.POST['q']
-            code = request.POST['code']
-            matches = []
-            if code == 'java':
-                src = os.path.join(settings.UPLD_DIR, md5+'/java_source/')
-                ext = '.java'
-            elif code == 'smali':
-                src = os.path.join(settings.UPLD_DIR, md5+'/smali_source/')
-                ext = '.smali'
-            else:
-                return HttpResponseRedirect('/error/')
-            # pylint: disable=unused-variable
-            # Needed by os.walk
-            for dir_name, sub_dir, files in os.walk(src):
-                for jfile in files:
-                    if jfile.endswith(ext):
-                        file_path = os.path.join(src, dir_name, jfile)
-                        if "+" in jfile:
-                            fp2 = os.path.join(src, dir_name, jfile.replace("+", "x"))
-                            shutil.move(file_path, fp2)
-                            file_path = fp2
-                        fileparam = file_path.replace(src, '')
-                        with io.open(
-                            file_path,
-                            mode='r',
-                            encoding="utf8",
-                            errors="ignore"
-                        ) as file_pointer:
-                            dat = file_pointer.read()
-                        if query in dat:
-                            matches.append(
-                                "<a href='../ViewSource/?file=" + escape(fileparam) +
-                                "&md5=" + md5 +
-                                "&type=apk'>" + escape(fileparam) + "</a>"
-                            )
+        md5_key = 'hash' if is_api else 'md5'
+        match = re.match('^[0-9a-f]{32}$', request.POST[md5_key])
+        include_path = request.POST.get('include_path', None)
+        if not match:
+            return HttpResponseNotFound()
+        md5 = request.POST[md5_key]
+        query = request.POST['q']
+        code = request.POST['code']
+        matches = []
+        if code == 'java':
+            src = os.path.join(settings.UPLD_DIR, md5+'/java_source/')
+            ext = '.java'
+        elif code == 'smali':
+            src = os.path.join(settings.UPLD_DIR, md5+'/smali_source/')
+            ext = '.smali'
+        else:
+            if is_api:
+                return HttpResponseBadRequest()
+            return HttpResponseRedirect('/error/')
+        # pylint: disable=unused-variable
+        # Needed by os.walk
+        for dir_name, sub_dir, files in os.walk(src):
+            if is_api and include_path and len(include_path) > 0:

```

Comment:
What is this code doing?

Reply:
Usually the entire directory will be scanned. This parameter can specify the directory.

Reply:
This is an optional parameter. The purpose of this parameter is to reduce the return of unnecessary information, such as only looking for the content under the current app, removing some third-party content.